Sprint 8 solution time and duration

LUKIDMI opened the pull request:
ФЗ 8 спринта

LUKIDMI committed:
В процессе выполнения ФЗ 8, добавлены поля duration, startTime и endTime

LUKIDMI committed:
Добавлен класс для преобразования задачи в строку и из строки в задачу

LUKIDMI committed:
Добавлен класс для конвертации задачи в строку для загрузки в файл и выгрузки из файла
Исправлены конструкторы задач
Добавлен метод, возвращающий задачи в порядке приоритета
Добавлены новые поля к задачам
Добавлен метод рассчета времени эпика
Исправлено много тестов и написаны новые
Добавлены ошибки....

LUKIDMI committed:
Добавлен пробел после конструктора Managers

LUKIDMI committed:
Убраны неиспользуемые импорты

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -1,37 +1,47 @@
 
+    //Удаляем все задачи
     @Override
     public void deleteAllTasks() {
```
нужно еще из prioritizedTasks удалить

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -103,17 +122,23 @@ public Epic getEpicById(int epicId) {
         epic.setId(id);
         epics.put(id, epic);
+
+        addPrioritizedTasks(epic);
```
эпики не хранятся в prioritizedTasks, они представлены там своими подзадачами

LUKIDMI replied:
исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -125,8 +150,40 @@ public void updateEpic(Epic epic) {
         existingEpic.setDescription(epic.getDescription());
 
         updateEpicStatus(existingEpic);
```
статус эпика не изменится, можно не делать

LUKIDMI replied:
Исправоено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -125,8 +150,40 @@ public void updateEpic(Epic epic) {
         existingEpic.setDescription(epic.getDescription());
 
         updateEpicStatus(existingEpic);
+        updateEpicTime(existingEpic);
```
аналогично

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -125,8 +150,40 @@ public void updateEpic(Epic epic) {
+        epic.setEndTime(endTime);
+
+        if (startTime != null && endTime != null) {
+            epic.setDuration(Duration.between(startTime, endTime));
```
продолжительность эпика - это сумма продолжительностей его подзадач, а не разница между startTime и endTime, потому что между подзадачами одного эпика, могут быть другие

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -139,73 +196,83 @@ private void updateEpicStatus(Epic epic) {
+                    historyManager.remove(subtaskId);
+                }
             }
-            historyManager.remove(epicId);
```
а зачем убрал удаление эпика из истории?

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -139,73 +196,83 @@ private void updateEpicStatus(Epic epic) {
+            historyManager.remove(subtask.getId());
         }
+
+        prioritizedTasks.removeAll(subTasks.values());
```
можно в предыдущем цикле for() поштучно subtask удалять из prioritizedTasks

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -217,52 +284,102 @@ public SubTask getSubTaskById(int subTaskId) {
         subTasks.put(id, subTask);
         epic.addSubTaskId(subTask);
+
+        updateEpicTime(epic);
```
еще нужно статус пересчитать

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -217,52 +284,102 @@ public SubTask getSubTaskById(int subTaskId) {
+
+        updateEpicTime(epic);
+
+        addPrioritizedTasks(subTask);
```
еще нужно из prioritizedTasks удалить старую версию задачи

LUKIDMI replied:
исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -217,52 +284,102 @@ public SubTask getSubTaskById(int subTaskId) {
             throw new IllegalArgumentException("Подзадача не найдена или равна null.");
         }
+
+        addPrioritizedTasks(subTask);
```
аналогично для подзадачи

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/model/Epic.java:
```diff
@@ -36,9 +49,46 @@ public void deleteSubTaskById(int subTaskId) {
         }
     }
 
+    public void setStartTime(LocalDateTime startTime) {
```
этот метод должен быть в родительском классе

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/model/Epic.java:
```diff
@@ -36,9 +49,46 @@ public void deleteSubTaskById(int subTaskId) {
+        this.endTime = endTime;
+    }
+
+    public void setDuration(Duration duration) {
```
и этот

LUKIDMI replied:
исправлено

irek-valiev commented on src/pavel/kalinkin/project/model/Epic.java:
```diff
@@ -36,9 +49,46 @@ public void deleteSubTaskById(int subTaskId) {
+    }
+
+    @Override
+    public Duration getDuration() {
```
для duration эпика расчет другой, а простой геттер берется из родительского класса

LUKIDMI replied:
исправлено

irek-valiev commented on src/pavel/kalinkin/project/model/Epic.java:
```diff
@@ -36,9 +49,46 @@ public void deleteSubTaskById(int subTaskId) {
+
+    @Override
+    public LocalDateTime getEndTime() {
+        if (endTime != null)
```
тут конструкция if/else лишняя, если endTime будет null, return endTime и return null одно и то же

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/model/SubTask.java:
```diff
@@ -1,27 +1,43 @@
+
 public class SubTask extends Task {
     private final int epicId;
 
```
нужно переопределить все родительские конструкторы

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/model/SubTask.java:
```diff
@@ -1,27 +1,43 @@
     }
 
+    @Override
+    public LocalDateTime getEndTime() {
```
для подзадач полностью подходит родительский метод, переопределять не нужно

LUKIDMI replied:
Исправлено

LUKIDMI committed:
Исправлены ошибки после ревью.
Переделан рассчет duration эпика
Добавлены конструкторы в подзадачи из родительского класса
Убраны геттеры/сеттеры родительского класса из классов наследников.

LUKIDMI committed:
Исправлены ошибки после ревью.
Переделан рассчет duration эпика
Добавлены конструкторы в подзадачи из родительского класса
Убраны геттеры/сеттеры родительского класса из классов наследников.

LUKIDMI committed:
Убран ненужный испорт

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -43,55 +55,64 @@ public Task getTaskById(int taskId) {
+
         tasks.put(task.getId(), task);
+
+        addPrioritizedTasks(task);
```
еще нужно из prioritizedTasks удалить старую версию задачи

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -139,73 +186,83 @@ private void updateEpicStatus(Epic epic) {
 
+    //Удаление эпика по id
     @Override
     public void deleteEpicById(int epicId) {
```
удалить эпик из истории

LUKIDMI replied:
исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -217,52 +274,103 @@ public SubTask getSubTaskById(int subTaskId) {
         subTask.setId(id);
         subTasks.put(id, subTask);
         epic.addSubTaskId(subTask);
+
```
пересчитать статус эпика

LUKIDMI replied:
Исправлено

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -217,52 +274,103 @@ public SubTask getSubTaskById(int subTaskId) {
         if (subTask == null || !subTasks.containsKey(subTask.getId())) {
             throw new IllegalArgumentException("Подзадача не найдена или равна null.");
         }
+
```
еще нужно из prioritizedTasks удалить старую версию подзадачи

LUKIDMI replied:
Исправлено

LUKIDMI committed:
Исправлены ошибки после ревью.
Добавлено удаление подзадачи из списка приоритетных после обновления

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -80,6 +80,7 @@ public void updateTask(Task task) {
 
         tasks.put(task.getId(), task);
 
+        prioritizedTasks.remove(task);
```
это новая версия, ее еще в prioritizedTasks  нет. 
нужно до обновления в хранилище tasks взять старую версию, удалить ее из prioritizedTasks, потом уже добавлять во все хранилища новую

LUKIDMI replied:
Готово

irek-valiev replied:
это все так же новая версия задачи. В prioritizedTasks.remove() нужно отдать старую версию задачи, которая лежит в tasks. А потом уже обновлять оба хранилища новой версией, которая task

LUKIDMI replied:
Поправил!)

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -305,6 +307,7 @@ public void updateSubTask(SubTask subTask) {
             throw new IllegalArgumentException("Подзадача не найдена или равна null.");
         }
 
+        prioritizedTasks.remove(subTask);
```
это тоже новая версия только подзадачи. Алгоритм действий такой же как с задачами

LUKIDMI replied:
Вроде бы ок теперь)

LUKIDMI committed:
исправлены ошибки после ревью

irek-valiev commented on src/pavel/kalinkin/project/manager/InMemoryTaskManager.java:
```diff
@@ -217,52 +274,106 @@ public SubTask getSubTaskById(int subTaskId) {
+        updateEpicStatus(epic);
+        updateSubTask(subTask);
+        updateEpicTime(epic);
+        prioritizedTasks.remove(subTask);
```
здесь все по аналогии с задачами. Нужно взять старую версию подзадачи из subTasks до того как в нее положить новую

LUKIDMI replied:
Не совсем понимаю алгоритм, наверно после перерыва. Если мы добавляем новую подзадачу, то ее же не должно быть prioritizedTasks?

LUKIDMI committed:
Исправлено после замечаний и длинного перерыва!)
